diff --git a/lib/backup/manager.rb b/lib/backup/manager.rb
index efaefa4..8f0e38f 100644
--- a/lib/backup/manager.rb
+++ b/lib/backup/manager.rb
@@ -7,7 +7,7 @@ module Backup
       s = {}
       s[:db_version]         = "#{ActiveRecord::Migrator.current_version}"
       s[:backup_created_at]  = Time.now
-      s[:gitlab_version]     = %x{git rev-parse HEAD}.gsub(/\n/,"")
+      s[:gitlab_version]     = File.read(Rails.root.join('REVISION')).chomp
       s[:tar_version]        = %x{tar --version | head -1}.gsub(/\n/,"")
 
       Dir.chdir(Gitlab.config.backup.path)
@@ -92,7 +92,7 @@ module Backup
         Dir.chdir(Rails.root)
 
         # restoring mismatching backups can lead to unexpected problems
-        if settings[:gitlab_version] != %x{git rev-parse HEAD}.gsub(/\n/, "")
+        if settings[:gitlab_version] != File.read(Rails.root.join('REVISION')).chomp
           puts "GitLab version mismatch:".red
           puts "  Your current HEAD differs from the HEAD in the backup!".red
           puts "  Please switch to the following revision and try again:".red
