diff --git a/setup.py b/setup.py
index 4b3c3f3..caf1b7e 100644
--- a/setup.py
+++ b/setup.py
@@ -2444,6 +2444,11 @@ class PyBuildExt(build_ext):
             ))
 
     def detect_nis(self):
+        # We do not want to include nis module as it links with system tirpc
+        # and thus fails healthcheck.
+        self.missing.append('nis')
+        return
+
         if MS_WINDOWS or CYGWIN or HOST_PLATFORM == 'qnx6':
             self.missing.append('nis')
             return
