diff --git a/unzip.c b/unzip.c
index 8dbfc95..954b204 100644
--- a/unzip.c
+++ b/unzip.c
@@ -570,8 +570,11 @@ Send bug reports using //www.info-zip.org/zip-bug.html; see README for details.\
 #else /* !VMS */
 # ifdef COPYRIGHT_CLEAN
    static ZCONST char Far UnzipUsageLine1[] = "\
-UnZip %d.%d%d%s of %s, by Info-ZIP.  Maintained by C. Spieler.  Send\n\
-bug reports using http://www.info-zip.org/zip-bug.html; see README for details.\
+UnZip %d.%d%d%s of %s, by GitLab Inc. The original UnZip sources are available\n\
+from Info-ZIP's home site at http://www.info-zip.org/pub/infozip/UnZip.html.\n\
+Our patches are available from\n\
+https://gitlab.com/gitlab-org/omnibus-gitlab/tree/master/config/patches/unzip.\n\
+Send any bug reports on this port to support@gitlab.com.\
 \n\n";
 # else
    static ZCONST char Far UnzipUsageLine1[] = "\
