From: Vincent Lefevre <vincent@vinc17.net>
Subject: Do not escape shell-special characters in "pat"
Bug-Debian: https://bugs.debian.org/1054628
X-Debian-version: 6.0-29

--- a/unix/zipgrep
+++ b/unix/zipgrep
@@ -49,10 +49,6 @@
 IFS='
 '
 
-# Escape shell-special characters in "pat".
-pat=` echo "$pat" | \
- sed -e 's/\\\\/\\\\\\\\/g' -e 's/|/\\\|/g' -e 's/&/\\\&/g' `
-
 # Use "unzip -Z1" to get a listing of the specified members from the
 # specified archive.  Escape any backslashes in a file name.
 for i in `unzip -Z1 "$zipfile" ${1+"$@"} | sed -e 's/\\\\/\\\\\\\\/g' `; do
