diff --git a/lib/chef/client.rb b/lib/chef/client.rb
index a0c3bbf331..626f0fac56 100644
--- a/lib/chef/client.rb
+++ b/lib/chef/client.rb
@@ -337,15 +337,6 @@ class Chef
 
     # @api private
     def warn_if_eol
-      require_relative "version"
-
-      # We make a release every year so take the version you're on + 2006 and you get
-      # the year it goes EOL
-      eol_year = 2006 + Gem::Version.new(Chef::VERSION).segments.first
-
-      if Time.now > Time.new(eol_year, 5, 01)
-        logger.warn("This release of #{ChefUtils::Dist::Infra::PRODUCT} became end of life (EOL) on May 1st #{eol_year}. Please update to a supported release to receive new features, bug fixes, and security updates.")
-      end
     end
 
     # @api private
