diff --git a/lib/chef/chef_fs/file_system/chef_server/cookbook_file.rb b/lib/chef/chef_fs/file_system/chef_server/cookbook_file.rb
--- a/lib/chef/chef_fs/file_system/chef_server/cookbook_file.rb      2021-01-13 21:18:02.000000000 -0800
+++ b/lib/chef/chef_fs/file_system/chef_server/cookbook_file.rb      2021-01-13 21:20:11.000000000 -0800
@@ -18,7 +18,7 @@

 require_relative "../base_fs_object"
 require_relative "../../../http/simple"
-require "openssl" unless defined?(OpenSSL)
+require "digest" unless defined?(Digest)

 class Chef
   module ChefFS
@@ -69,7 +69,7 @@
           private

           def calc_checksum(value)
-            OpenSSL::Digest::MD5.hexdigest(value)
+            ::Digest::MD5.hexdigest(value)
           end
         end
       end
