diff --git a/Makefile b/Makefile
index bb3f488bee..662387bc7d 100644
--- a/Makefile
+++ b/Makefile
@@ -221,7 +221,7 @@ list-tests:
 	 $(PERL) $(SRCDIR)/test/run_tests.pl list
 	@ : 
 
-install: install_sw install_ssldirs install_docs
+install: install_sw install_ssldirs
 
 uninstall: uninstall_docs uninstall_sw
 
@@ -269,8 +269,6 @@ install_sw: install_dev install_engines install_runtime
 
 uninstall_sw: uninstall_runtime uninstall_engines uninstall_dev
 
-install_docs: install_man_docs install_html_docs
-
 uninstall_docs: uninstall_man_docs uninstall_html_docs
 	$(RM) -r -v $(DESTDIR)$(DOCDIR)
 
@@ -495,24 +493,12 @@ uninstall_runtime_libs:
 	@ : 
 
 
-install_man_docs:
-	@[ -n "$(INSTALLTOP)" ] || (echo INSTALLTOP should not be empty; exit 1)
-	@$(ECHO) "*** Installing manpages"
-	$(PERL) $(SRCDIR)/util/process_docs.pl \
-		--destdir=$(DESTDIR)$(MANDIR) --type=man --suffix=$(MANSUFFIX)
-
 uninstall_man_docs:
 	@$(ECHO) "*** Uninstalling manpages"
 	$(PERL) $(SRCDIR)/util/process_docs.pl \
 		--destdir=$(DESTDIR)$(MANDIR) --type=man --suffix=$(MANSUFFIX) \
 		--remove
 
-install_html_docs:
-	@[ -n "$(INSTALLTOP)" ] || (echo INSTALLTOP should not be empty; exit 1)
-	@$(ECHO) "*** Installing HTML manpages"
-	$(PERL) $(SRCDIR)/util/process_docs.pl \
-		--destdir=$(DESTDIR)$(HTMLDIR) --type=html
-
 uninstall_html_docs:
 	@$(ECHO) "*** Uninstalling manpages"
 	$(PERL) $(SRCDIR)/util/process_docs.pl \
-- 
2.17.2 (Apple Git-113)

