diff --git a/GNUmakefile.in b/GNUmakefile.in
index 2352fc1..0191427 100644
--- a/GNUmakefile.in
+++ b/GNUmakefile.in
@@ -13,7 +13,7 @@ $(call recurse,all install,src config)
 docs:
 	$(MAKE) -C doc all

-$(call recurse,world,doc src config contrib,all)
+$(call recurse,world,src config contrib,all)

 # build src/ before contrib/
 world-contrib-recurse: world-src-recurse
@@ -29,7 +29,7 @@ html man:
 install-docs:
 	$(MAKE) -C doc install

-$(call recurse,install-world,doc src config contrib,install)
+$(call recurse,install-world,src config contrib,install)

 # build src/ before contrib/
 install-world-contrib-recurse: install-world-src-recurse
